### PR TITLE
Support exclude patterns

### PR DESCRIPTION
**Pull request by jherland**
Support gitignore-style patterns given directly in the configuration or on the command line. No support for indirectly reading ignore patterns from files, yet.

See #391 for a follow-up PR that vendors/integrates the `gitignore_parser` dependency instead of depending on it, as we cannot really use its "public" API as intended.

Commits:
- `test_dir_traversal`: Add `.skip_me` member to each `DirectoryTraversalVector`
- Add `gitignore_parser` as a new dependency
- `DirectoryTraversal`: Initial support for exclude patterns
- `dir_traversal`: Add `excluded_*` fields to `TraversalStep`
- `dir_traversal`: Do not exclude an explicitly `.add()`ed path
- `traverse_project`: Use new `exclude` feature to exclude "hidden" paths
- `Settings`: Add `.exclude`, paths to exclude during project traversal
- `sample_projects`: Add support for exclude patterns
- Add `--exclude` option to CLI
- Add `ExcludePattern` test helper

---

(Old description before rename from "ignore" to "exclude":)

Support gitignore-style patterns given directly in the configuration or on the command line. No support for _files_ with ignore patterns yet.

Also, contemplating whether to vendor/integrate the `gitignore_parser` dependency instead of depending on it, as we cannot really use its "public" API as intended.

Commits:
- Add `gitignore_parser` as a new dependency
- `DirectoryTraversal`: Initial support for ignore patterns
- `dir_traversal`: Add `ignored_*` fields to `TraversalStep`
- `dir_traversal`: Do not ignore an explicitly `.add()`ed path
- `traverse_project`: Use new ignore feature to ignore "hidden" paths
- `Settings`: Add `.ignores`, paths to ignore during project traversal
- `sample_projects`: Add support for ignores
- Add `--ignore` option to CLI

### CODE REVIEW

File: tests/test_dir_traversal.py
Comment (Nour-Mws) [inline tests/test_dir_traversal.py:316]: I am not sure I follow why `", "` is a blank line

File: tests/test_traverse_project.py
Comment (Nour-Mws) [inline tests/test_traverse_project.py:526]: The error is caused by the tab on the second line? or the general lack of ignore patterns?

File: tests/test_traverse_project.py
Comment (Nour-Mws) [inline tests/test_traverse_project.py:542]: How would this not raise the `IgnoreRuleError`?

File: tests/sample_projects/hidden_files/expected.toml
Comment (Nour-Mws) [inline tests/sample_projects/hidden_files/expected.toml:11]:
```suggestion
description = "Default run where everything is hidden."
```

File: tests/sample_projects/hidden_files/expected.toml
Comment (Nour-Mws) [inline tests/sample_projects/hidden_files/expected.toml:28]: Why aren't the pyenvs under `.venvs/`detected if the dot ignore pattern is overridden?

File: tests/sample_projects/hidden_files/expected.toml
Comment (Nour-Mws) [inline tests/sample_projects/hidden_files/expected.toml:57]: Same question as above

File: README.md
Comment (Nour-Mws) [inline README.md:318]: Which example are we referring to?

**Review by Nour-Mws — COMMENTED** [review]
Meticulous and comprehensive as expected :sparkles: 
I am a bit confused as to some tests related to ignored venvs and I'm sure this is a lapse on my side as the tests are passing.
But I thought I'd comment and wait until I understand so that I can approve in good conscience.

File: README.md
Comment (jherland) [reply README.md:318]: `fawltydeps --ignore tests/` on line 307

File: tests/sample_projects/hidden_files/expected.toml
Comment (jherland) [reply tests/sample_projects/hidden_files/expected.toml:28]: Oh, crap. This is poorly documented...

The `pyenvs = [""]` here is not the Python environments that we expect to find (i.e. the expected output), but rather the `--pyenv` values that we pass _into_ the analysis (that is, we're passing `""` which is appended onto the directory containing the sample project, meaning that we're searching through the entire sample project for Python envs). The Python environments that we end up finding are actually _not_ directly verified here at all...

File: tests/test_traverse_project.py
Comment (jherland) [reply tests/test_traverse_project.py:526]: The general lack of ignore patterns. The intention here was to test that a blank line is considered empty whether or not it contains extra whitespace. These should be three different tests, I think.

File: tests/test_traverse_project.py
Comment (jherland) [reply tests/test_traverse_project.py:542]: In order to make `"foo/bar"` a valid pattern in this context, we would have to specify a `base_dir` (required when the pattern is "anchored", i.e. when it contains a slash). You can see tests of this logic in `test_dir_traversal.py`, specifically look at `anchored_pattern__must_have_base_dir` and `ignore_pattern_without_slash__does_not_match_above_base_dir` for examples without/with `base_dir` respectively.

But this logic is not _yet_ available to users of `traverse_project`. At this point in the series of PRs, we can only specify ignore patterns in the CLI, and from the CLI I believe it is _unclear_ what the correct `base_dir` should be: The natural choice would be the current dir, but as soon as you pass multiple `basepath`s on the command line, and/or multiple `--code` or `--deps` paths, I think it becomes impossible to determine which `base_dir` to be used for `--ignore foo/bar`. Therefore, I've chosen to leave `base_dir` unspecified in this case, with the consequence that anchored patterns cannot be given directly to `--ignore` on the command-line.

This is reflected in our code as of this PR: `DirectoryTraversal.ignore()` does take an optional `base_dir`, but this argument is not yet used in `traverse_project.find_sources()`, and it will only start to be used when I add the logic for parsing `.gitignore` files (in the final `--ignore-from` PR that I'm still working on). When parsing `.gitignore` files, the `base_dir` is indeed well-specified: IIRC, [the `gitignore` documentation](https://git-scm.com/docs/gitignore) states that patterns in a file use the parent directory of that file as their `base_dir`.

File: tests/test_dir_traversal.py
Comment (jherland) [reply tests/test_dir_traversal.py:316]: There are two blank lines: `""`, and `""`...

File: tests/test_dir_traversal.py
Comment (jherland) [reply tests/test_dir_traversal.py:317]: We're checking that we get `IgnoreRuleMissing` raised twice, once for each empty pattern above.

[In the test below](https://github.com/tweag/FawltyDeps/pull/388/files/142b91a7559c389414ac807556daeff5d3919074#diff-3aa986ddb8b14d40d3559e3b54826bfc5b43b56d33d849bd21f82378c8723764R763) we pass each ignore pattern separately to the parser and recording each exception raised. I wanted to make sure that the first exception is not messing up some internal state that would cause problems for later ignore patterns. (When we get around to parsing gitignore files, we will encounter valid ignore patterns interspersed with empty lines, and we need to skip the empty lines and keep parsing.)

File: tests/test_dir_traversal.py
Comment (jherland) [reply tests/test_dir_traversal.py:324]: Here is everything that the [gitignore docs](https://git-scm.com/docs/gitignore) have to say about backslashes:
> A line starting with # serves as a comment. Put a backslash ("`\`") in front of the first hash for patterns that begin with a hash.
>
> Trailing spaces are ignored unless they are quoted with backslash ("`\`").
>
> An optional prefix "`!`" which negates the pattern; any matching file excluded by a previous pattern will become included again. It is not possible to re-include a file if a parent directory of that file is excluded. Git doesn’t list excluded directories for performance reasons, so any patterns on contained files have no effect, no matter where they are defined. Put a backslash ("`\`") in front of the first "`!`" for patterns that begin with a literal "`!`", for example, "`\!important!.txt`".

As for hash characters elsewhere on the line, I just tested, and these are _not_ interpreted as comment chars, but rather as part of the pattern (so no backslash needed there). This is in line with the above docs, but maybe unexpected when comparing to another languages like e.g. shell or Python.

File: tests/test_dir_traversal.py
Comment (jherland) [reply tests/test_dir_traversal.py]: Added test helper in [2f13434](https://github.com/tweag/FawltyDeps/pull/388/commits/2f13434038593f3a5a070f98f6a5a625e689a15d).

File: tests/test_dir_traversal.py
Comment (jherland) [reply tests/test_dir_traversal.py:513]: Sorry, this was my horribly under-documented test code striking again... The `a/` is the base_dir for the `a/b/` pattern, not a pattern in its own right...

Hopefully clearer after 2f13434.

File: tests/test_dir_traversal.py
Comment (jherland) [reply tests/test_dir_traversal.py]: Almost. Technically speaking `abc/` ignores the `abc` directory, but `abc/**` ignores everything _inside_ that directory. From the point of view of the `DirectoryTraversal` itself, the `abc/` dir _is_ traversed (but nothing inside it). You can see this in the expected traversal step below (the second one):

```python
            ExpectedTraverseStep("abc"),  # ignored_subdirs=["a"], ignored_files=["def"]
```

This step would not occur if we ignored the `abc/` directory instead.

In practice, from the FawltyDeps traversal's POV, the difference is academic. I'm trying to construct a case where there would be a meaningful difference in behavior, but it's hard to come up with: For 'code' and 'deps', we're looking for files, so the extra directory traversal step makes no difference. For `pyenv` we are only ignoring for Python environments that are located _inside_ ignored dirs (i.e. with `.*` as the default pattern we _want_ to find `.venv/`, but we want to ignore `.hidden/.venv/`). Therefore, I believe a virtualenv located at `abc/my_venv` would be _ignored_ by the `abc/` pattern, but would be _found_ by the `abc/**` pattern.

File: tests/test_dir_traversal.py
Comment (jherland) [reply tests/test_dir_traversal.py]: Yes, you got it! As I've now tried to fix in 2f13434, the tuple was used when we needed to communicate an ignore pattern together with a `base_dir` (in practice: the directory of the `.gitiignore` file containing this pattern). Wrapping this in an `IgnorePattern("**/foo", base_dir="a/")` hopefully makes it more readable.

**Comment by jherland** [comment]
> Meticulous and comprehensive as expected ✨
> I am a bit confused as to some tests related to ignored venvs and I'm sure this is a lapse on my side as the tests are passing.
> But I thought I'd comment and wait until I understand so that I can approve in good conscience.

Thanks for the review, and I apologize for not getting around to replying sooner.

Yes, I stupidly used `List[Union[str, Tuple[str, Optional[str]]]]` for the `DirectoryTraversalVector.ignore_patterns` member, which only added unnecessary confusion to something that is already quite complex. 2f13434 tries to improve on that.

Also, in a couple of my replies above I've tried to explain the "anchored" patterns that require the `base_dir` argument to `DirectoryTraversal.ignore()`. I regret adding (and testing) this extra-complex part here in this PR, when it is _not_ used from the POV of `traverse_project`. The reason we're not properly using this part yet, is because I don't think there _is_ a suitable `base_dir` that can be set for patterns that come from the command line. They can only be determined for patterns that come from `.gitignore` files, and support for those are only added in a future PR...

Again, I'm sorry for the unnecessary complexity that is not immediately warranted by this PR. Maybe I should remove the `base_dir` argument from `DirectoryTraversal.ignore()` for now? Or at least add a comment explaining how/why it's not yet used by `traverse_project`?

File: tests/sample_projects/hidden_files/expected.toml
Comment (Nour-Mws) [reply tests/sample_projects/hidden_files/expected.toml:28]: Oops

File: tests/test_dir_traversal.py
Comment (Nour-Mws) [reply tests/test_dir_traversal.py]: Thanks for the explanation!

**Review by Nour-Mws — APPROVED** [review]
Thanks for taking the time to explain in details. I have a much better understanding of the context now and the blind spots in the tests are clearer.
It took me more than I expected to go through the replies now that a certain duration has passed since I've last reviewed this.
All good. This is ready to go!

**Comment by jherland** [comment]
@mknorps: I had to make a few changes here to reconcile this PR with your recent Windows work. In short, I added commit 3a2b96f to refactor how we skip individual `DirectoryTraversalVector`s on Windows, and I additionally made the following changes (compared to what was already approved) to make the tests pass on Windows and MacOS:

```diff
diff --git a/tests/test_cmdline.py b/tests/test_cmdline.py
index 196d31f..1ba9bb2 100644
--- a/tests/test_cmdline.py
+++ b/tests/test_cmdline.py
@@ -241,7 +241,7 @@ def test_list_imports__from_dir_with_some_ignored__prints_imports_from_unignored
         }
     )
 
-    expect = [f"{tmp_path}/file3.ipynb[1]:1: pytorch"]
+    expect = [f"{tmp_path / 'file3.ipynb'}[1]:1: pytorch"]
     output, returncode = run_fawltydeps_function(
         "--list-imports", "--detailed", f"--code={tmp_path}", "--ignore=*.py"
     )
diff --git a/tests/test_dir_traversal.py b/tests/test_dir_traversal.py
index e3e9153..fef53b2 100644
--- a/tests/test_dir_traversal.py
+++ b/tests/test_dir_traversal.py
@@ -375,6 +387,8 @@ directory_traversal_vectors: List[DirectoryTraversalVector] = [
                 ".", files=["do_not_ignore "], ignored_files=["do_ignore "]
             ),
         ],
+        skip_me=on_windows("Windows does not support paths with trailing spaces"),
+        # https://stackoverflow.com/a/67886889
     ),
     # An optional prefix "!" which negates the pattern; any matching file
     # excluded by a previous pattern will become included again. It is not
@@ -608,7 +622,7 @@ directory_traversal_vectors: List[DirectoryTraversalVector] = [
         "range_matches_any_one_char_in_range",
         given=[
             File("abcdef"),  # not matched
-            File("abcxdef"),  # not matched
+            File("abcWdef"),  # not matched
             File("abcXdef"),
             File("abcYdef"),
             File("abcZdef"),
@@ -620,7 +634,7 @@ directory_traversal_vectors: List[DirectoryTraversalVector] = [
             ExpectedTraverseStep(
                 ".",
                 subdirs=["abc"],
-                files=["abcdef", "abcxdef", "abcXYXdef"],
+                files=["abcdef", "abcWdef", "abcXYXdef"],
                 ignored_files=["abcXdef", "abcYdef", "abcZdef"],
             ),
             ExpectedTraverseStep("abc", files=["def"]),
```

File: fawltydeps/settings.py
Comment (GuillaumeDesforges) [inline fawltydeps/settings.py:165]: Since Maria mentionned it in the dataeng weekly, I'll just add a quick comment that can be safely ignored.

Since FawltyDeps "ignore" can be about different things, namely import of file paths, as a user I may be confused by the name `--ignores`. A more explicit option name (e.g. `--ignore-files`) would be less confusing to me.

Comment (mknorps) [reply fawltydeps/settings.py:165]: We have 
- `--ignore-undeclared` to ignore imports that do not have declared dependencies
- `--ignore` for paths we traverse (including files).
So we have an explicit name but for a first case you mentioned. Is this OK?

File: tests/test_dir_traversal.py
Comment (mknorps) [inline tests/test_dir_traversal.py:136]: Maybe we could simplify the type here to just:
```suggestion
    skip_me: Optional[str] = None
```
We use only `on_windows` function to pupolate this field for now, and maybe we do not need to over-abstarct this.
Unless there is some other reason, that skips me.

**Review by mknorps — APPROVED** [review]
I imagine this was a big rebase :1st_place_medal: ...

It looks good to me. I had one question/suggestion regarding the `skip_me` option.

File: fawltydeps/settings.py
Comment (jherland) [reply fawltydeps/settings.py:165]: Thanks for raising this question @GuillaumeDesforges. Although we are eager to get this new feature merged and published, this is probably the best time to make sure we get the naming right  (_before_ committing to it in a public release).

FTR, we also have:
- `--ignore-unused` to ignore dependencies that are not (meant to be) imported, such as tools, etc.
- `--ignore-from`: This does not exist yet, but it's what I'm currently working on to support getting `--ignore` pattens indirectly from a file of patterns. To be used like this: `fawltydeps --ignore-from=.gitignore`.

Although all of these options have different names, I cannot prove that this will **NOT** be confusing to some users. There is nothing inherent in the name of  `--ignore` (or `--ignore-from`) that explains that these are about ignoring _paths_. Instead, we hope/assume/believe that users will most easily associate the concept of ignoring _something_ to the paths that are given as input to FawltyDeps. (And if not, we hope that `--help` or the docs will quickly set them straight.)

We could go with something longer, like `--ignore-paths` (and `--ignore-paths-from`?), which would be clearer, but more tedious to type out.

We could also try to re-evaluate our naming and come up with a better scheme. For one: `--ignore-undeclared` and `--ignore-unused` are about ignoring facets of the analysis done by FawltyDeps, but ignoring paths is really about excluding paths from the analysis entirely. As such the path-related options are logically applied/handled at a very different time than the other two. We could consider something like `--exclude`/`--exclude-from` to denote that these are _different_ from the existing `--ignore-undeclared`/`--ignore-unused`? That said, I still find the overlapping of terms across different tools a little clunky here: i.e. that the "gitignore option" is named `--exclude-from=.gitignore` when we have other options called `--ignore-something`.

Maybe we should take a closer look at how other linter tools handle this situation, although I suspect we won't find much consensus...

File: tests/test_dir_traversal.py
Comment (jherland) [reply tests/test_dir_traversal.py:136]: True, I was thinking _too_ generically here (that the test vector should be able to check something _when the test is run_ to decide whether or not to skip), but (since Python is an interpreted language) the Windows-or-not question is already known when we create the test vector here. Will fix. 🙏🏻

Comment (jherland) [reply tests/test_dir_traversal.py]: Fixed in 584fc1d45a1b570fd414e74b4b8cf3539674978c.

File: fawltydeps/settings.py
Comment (mknorps) [reply fawltydeps/settings.py:165]: I really like this `--exclude`/`--exclude-from` idea!
It clearly separates what we ignore in traversing directories and what we ignore when showing results. 
In other words `--exclude...` is in the preprocessing step and `--ignore...` in post-processing.

Currently we have one CLI option that uses file as a helper for a setup and it is `--custom-mapping-file`. `--exclude-file` would be extremely confusing, since its purpose it to load a file that defines ignored paths in the gitignore-style. Maybe we should consider having `--custom-mapping-from`. But this would mean backward incompatibility.

Comment (jherland) [reply fawltydeps/settings.py:165]: A quick survey of what other linters do:
- ruff:
  - [`exclude`](https://docs.astral.sh/ruff/settings/#exclude) works on "file patterns" (i.e. adjust analysis input)
  - [`ignore`](https://docs.astral.sh/ruff/settings/#ignore) works on "rule codes" (i.e. adjust result of analysis)
- flake8:
  - [`--exclude`](https://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-exclude) works on files.
  - [`--ignore`](https://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-ignore) works on "codes".
- pylint:
  - [`--ignore` (and friends)](https://pylint.pycqa.org/en/latest/user_guide/configuration/all-options.html#ignore) are used to ignore paths.
  - [`--disable`](https://pylint.pycqa.org/en/latest/user_guide/configuration/all-options.html#disable) seems to be the main way to disable checkers/rules/codes.
  - Various checkers add other `--ignore-*` options, but these are for ignoring checker-specific details (not related to paths at all).
  - A couple of checkers add `--exclude-*` options, for similar purposes.
  - In general, I don't think pylint is what we should try to emulate. It's interface is a bewildering mess, IMHO.
- black:
  - [`--exclude`](https://black.readthedocs.io/en/stable/usage_and_configuration/the_basics.html#exclude): "A regular expression that matches files and directories that should be excluded on recursive searches."
  - Black is famously lacking in options to change its logic/behavior, but there are a few options named [`--skip-*`](https://black.readthedocs.io/en/stable/usage_and_configuration/the_basics.html#x-skip-source-first-line) to disable a few parts of its internal logic.
- Mypy:
  - [`exclude`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-exclude) is used to exclude paths
  - For ignoring/disabling internal rules, there is a plethora of options with names like `disallow_*` or `warn_*`.
  - There is not much use of `ignore`, but the few uses that are visible are certainly more related to rules ([`ignore_errors`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-ignore_errors), [`ignore_missing_imports`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-ignore_missing_imports), [`warn_unused_ignores`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-warn_unused_ignores), or [`# type: ignore`](https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html#silencing-type-errors)) than to paths/files.

I did not expect to find this much consensus, and it seems a change of opinion is in order:

I now think we should rename `--ignore` (and `--ignore-from`) to `--exclude` (and `--exclude-from`[^1]).

What do you think, @GuillaumeDesforges and @mknorps?

[^1]: Although it's worth having another look at how other tools deal with ignore patterns read from a file like `.gitignore`.

Comment (jherland) [reply fawltydeps/settings.py:165]: (I didn't see your last comment @mknorps before I posted the above survey, but it seems we're in agreement. 🎉)

Re. renaming `--custom-mapping-file` to `--custom-mapping-from`: Yes, it's backwards incompatible, so we would have to keep `...-file` as a deprecated alias to `...-from` for a while, before we remove it completely. Other than that, I don't think it's a problem to rename this.

Comment (mknorps) [reply fawltydeps/settings.py:165]: > I now think we should rename --ignore (and --ignore-from) to --exclude (and --exclude-from[1](https://github.com/tweag/FawltyDeps/pull/388#user-content-fn-1-67c636c8d29ad3ceedc5aad716e06f53)).

aye!

Comment (GuillaumeDesforges) [reply fawltydeps/settings.py:165]: Makes sense to me :+1: this is probably what triggered my initial comment, somehow my brain had internalized how other tools behave :D

Comment (jherland) [reply fawltydeps/settings.py]: Thanks @GuillaumeDesforges for bothering us enough to fix it before we released it! 😅

I've rebased/replaced everything from `ignore` to `exclude` now. I took extra care in the docs to make sure things are as clear and unambiguous as I could make them.

**Review by mknorps — CHANGES_REQUESTED** [review]
There are some surprising behaviors with `--exclude` option. It looks like the problem is related to absolute vs relative paths, but it should be checked.

Two examples below:

- Example 1

    Run FawltyDeps on itself with `--exclude` option like that:
    ```
    (fawltydeps-RB6LwJVc-py3.11) ➜  FawltyDeps git:(jherland/ignore-work) fawltydeps --exclude fawltydeps
    
    No undeclared or unused dependencies detected.
    ```

    I thought I will have lots of unused dependencies.

- Example 2
  ```
  (fawltydeps-RB6LwJVc-py3.11) ➜  FawltyDeps git:(jherland/ignore-work) fawltydeps . --exclude 'fawltydeps/traverse_project.py'
  usage: fawltydeps [--check | --check-undeclared | --check-unused | --list-sources | --list-imports | --list-deps]
                [--summary | --detailed | --json] [--code PATH_OR_STDIN [PATH_OR_STDIN ...]]
                [--deps PATH [PATH ...]]
                [--deps-parser-choice {requirements.txt,setup.py,setup.cfg,pyproject.toml}]
                [--pyenv PYENV_DIR [PYENV_DIR ...]] [--exclude PATTERN [PATTERN ...]] [--install-deps]
                [--custom-mapping-file FILE_PATH [FILE_PATH ...]] [--config-file CONFIG_FILE]
                [--ignore-undeclared IMPORT_NAME [IMPORT_NAME ...]] [--ignore-unused DEP_NAME [DEP_NAME ...]]
                [-v] [-q] [--generate-toml-config] [-V] [-h]
                [basepaths ...]
  fawltydeps: error: Error while parsing exclude pattern: Anchored pattern without base_dir in 'fawltydeps/traverse_project.py'
  ```

**Comment by jherland** [comment]
> * Example 1
>   Run FawltyDeps on itself with `--exclude` option like that:
>   ```
>   (fawltydeps-RB6LwJVc-py3.11) ➜  FawltyDeps git:(jherland/ignore-work) fawltydeps --exclude fawltydeps
>   
>   No undeclared or unused dependencies detected.
>   ```
>   
>   I thought I will have lots of unused dependencies.

There are certainly problems with absolute vs relative paths that are still to be resolved. But this example  has _another_ complicating factor as well: Our own configuration from pyproject.toml contains: `code = ["fawltydeps"]`, meaning that we end up with both `code = {"fawltydeps"}` and `exclude = {"fawltydeps"}` in our `Settings` object, and in that case, the `code` takes precedence over the `exclude`. This was not obvious to you, the user. I'll certainly add a warning about this special case.

**Comment by jherland** [comment]
> * Example 2
>   ```
>   $ fawltydeps . --exclude 'fawltydeps/traverse_project.py'
>   usage: [...]
>   fawltydeps: error: Error while parsing exclude pattern: Anchored pattern without base_dir in 'fawltydeps/traverse_project.py'
>   ```

This is currently expected (but unfortunate). The base_dir referred to in the error message is not the same as a the basepath you supplied on the command line, but rather a separate piece of data that follows the exclude pattern: When the pattern is read from a `.gitignore` file, the base_dir is the directory holding this `.gitignore` file. (and it means that the pattern will only apply to paths _within_ this directory, e.g. it prevents a pattern in `foo/.gitignore` to be applied to paths outside `foo/`). The base_dir is also needed whenever the pattern is "anchored"[^1]. However, this not so easy when the pattern does not come from a `.gitignore` file, but comes from the command-line instead.

Here are some examples to show why this does not easily work:
- `fawltydeps --exclude foo/bar`: Here the default basepath is `./` so we could easily argue that the `foo/bar` pattern must match `./foo/bar` only.
- `fawltydeps path/to/my_project --exclude foo/bar`: Now we've changed the basepath. We _could_ argue that the pattern should be interpreted relative to the customized basepath, i.e. the `foo/bar` pattern should match `path/to/my_project/foo/bar` only.
- `fawltydeps path/to/my/scripts path/to/my/notebooks --exclude foo/bar`: Now we have _two_ basepaths. Where does the pattern apply?
  - Both `path/to/my/scripts/foo/bar` and `path/to/my/notebooks/foo/bar`?
  - None of them?
  - Auto-calculate the common parent and use that: `path/to/my/foo/bar`? Probably not...
- `fawltydeps --code path/to/my/scripts --deps path/to/my/deps --exclude foo/bar`: Now things are getting really complex: We're looking for code under `path/to/my/scripts`, deps under `path/to/my/deps`, and Python environments under `./`. But where should the pattern apply?
  - All places? `path/to/my/scripts/foo/bar`, `path/to/my/deps/foo/bar`, and `./foo/bar`?
  - None of them?
  - Something else?

I decided early on to NOT deal with this problem just yet, and thus the current behavior is that CLI-supplied patterns get no `base_dir`, and thus cannot be anchored (i.e. the only place you can put a `/` is at the end, to signal a directory-only match). But it is becoming clear to me that we rather need to fix this now, _before_ I have finished the third PR...

Right now, I am leaning towards using all paths as base_dirs, that is, for the last example above, the `foo/bar` exclude pattern technically ends up as three separate exclude patterns (with `path/to/my/scripts`, `path/to/my/deps`, and `./` as their respective base_dirs). At least I hope this would cause the least amount of surprises...

[^1]:  This comes from the `git help ignore` "specification": A pattern with a `/` at the start or in the middle is always interpreted relative to the directory in which the pattern was found.

**Comment by jherland** [comment]
For the record, I'm fixing the above issues in a new PR: #415. Although I'm not rewriting much code, and could potentially squash the fixes into this PR, I would have to rebase them across PR #391, and that introduces a substantial amount of conflicts that I'd rather avoid. This PR and #391 are already reviewed and with #415 fixing these latest issues, I plan on merging all 3 to `main` once #415 is finished and reviewed.

**Comment by jherland** [comment]
#415 is ready for review now. It should fix the issues you found above, @mknorps.

**Review by mknorps — APPROVED** [review]
With #415 this is ready to be merged :)

**Pull request merged**
